contracts/contracts/token/WOETH.sol

```diff
@@ -49,7 +90,84 @@ contract WOETH is ERC4626, Governable, Initializable {
         require(asset_ != address(asset()), "Cannot collect OETH");
         IERC20(asset_).safeTransfer(governor(), amount_);
     }
+
+    /**
+     * @dev This function converts requested OETH token amount to its underlying OETH
+     * credits value that is stored internally in OETH.sol and is required in order to
+     * be able to rebase.
+     *
+     * @param oethAmount Amount of OETH to be converted to OETH credits
+     * @return amount of OETH credits the OETH amount corresponds to
+     */
+    function _oethToCredits(uint256 oethAmount) internal returns (uint256) {
+        (, uint256 creditsPerTokenHighres, ) = OETH(asset())
+            .creditsBalanceOfHighres(address(this));
+
+        /**
+         * Multiplying OETH amount with the creditsPerTokenHighres is exactly the math that
+         * is internally being done in OETH:
+         */
+        // solhint-disable-next-line max-line-length
+        /** https://github.com/OriginProtocol/origin-dollar/blob/2314cccf2933f5c1f76a6549c1f5c9cc935b6f05/contracts/contracts/token/OUSD.sol#L242-L249
+         *
+         * This should make sure that the rounding will always be correct / mimic the rounding
+         * of OETH.
+         */
+        return oethAmount.mulTruncate(creditsPerTokenHighres);
+    }
+
+    /** @dev See {IERC4262-totalAssets} */
+    function totalAssets() public view virtual override returns (uint256) {
+        (, uint256 creditsPerTokenHighres, ) = OETH(asset())
+            .creditsBalanceOfHighres(address(this));
+
+        return (oethCreditsHighres).divPrecisely(creditsPerTokenHighres);
+    }
+
+    /** @dev See {IERC4262-deposit} */
+    function deposit(uint256 oethAmount, address receiver)
+        public
+        virtual
+        override
+        returns (uint256 woethAmount)
+    {
+        woethAmount = super.deposit(oethAmount, receiver);
+        oethCreditsHighres += _oethToCredits(oethAmount);
+    }
+
+    /** @dev See {IERC4262-mint} */
+    function mint(uint256 woethAmount, address receiver)
+        public
+        virtual
+        override
+        returns (uint256 oethAmount)
+    {
+        oethAmount = super.mint(woethAmount, receiver);
+        oethCreditsHighres += _oethToCredits(oethAmount);
+    }
+
+    /** @dev See {IERC4262-withdraw} */
+    function withdraw(
+        uint256 oethAmount,
+        address receiver,
+        address owner
+    ) public virtual override returns (uint256 woethAmount) {
+        woethAmount = super.withdraw(oethAmount, receiver, owner);
+        oethCreditsHighres -= _oethToCredits(oethAmount);
+    }
+
+    /** @dev See {IERC4262-redeem} */
+    function redeem(
+        uint256 woethAmount,
+        address receiver,
+        address owner
+    ) public virtual override returns (uint256 oethAmount) {
+        oethAmount = super.redeem(woethAmount, receiver, owner);
+        oethCreditsHighres -= _oethToCredits(oethAmount);
```

Comment:
This is just a mathematical nit, and perhaps the code is okay without this, but in general, if you are depositing/plus-ing and withdrawing/minus-ing using the same conversion function, you are almost certainly rounding the wrong direction in one of them. It's possible we need two `_oethToCredits`, one that round up, and one that rounds down.